### PR TITLE
Run product senders in parallel by default

### PR DESCRIPTION
**Pull request by jmfee-usgs**

### CODE REVIEW

**Review by emartinez-usgs — CHANGES_REQUESTED** [review]
Maybe nothing really needs to change, but I did ask some questions that would put my mind at ease if answered prior to merging.

File: src/main/java/gov/usgs/earthquake/distribution/CLIProductBuilder.java [hunk 1]
```diff
@@ -239,6 +250,51 @@ public void startup() throws Exception {
...
+					sendExceptions.put(sender, e);
+				}
+				return null;
+			});
+		}
+		// run in parallel
+		ExecutorService sendExecutor = Executors.newFixedThreadPool(senders.size());
+		try {
+			sendExecutor.invokeAll(sendTasks, timeoutSeconds, TimeUnit.SECONDS);
+		} catch (InterruptedException e) {
```
[the remark this answers is not in the document]
Comment (jmfee-usgs) [reply src/main/java/gov/usgs/earthquake/distribution/CLIProductBuilder.java:285]: I think this is if the process is terminated before completing (ctrl+c).

File: src/main/java/gov/usgs/earthquake/distribution/CLIProductBuilder.java [hunk 2]
```diff
@@ -239,6 +250,51 @@ public void startup() throws Exception {
...
+					sendComplete.put(sender, true);
+				} catch (Exception e) {
+					sendExceptions.put(sender, e);
+				}
+				return null;
+			});
+		}
+		// run in parallel
+		ExecutorService sendExecutor = Executors.newFixedThreadPool(senders.size());
+		try {
```
[the remark this answers is not in the document]
Comment (jmfee-usgs) [reply src/main/java/gov/usgs/earthquake/distribution/CLIProductBuilder.java:283]: It likely fails.  Not sure where NPE would come from different from regular send.  RejectedExecution is also unexpected, but I could add a try to fall back to normal send?

File: src/main/java/gov/usgs/earthquake/distribution/CLIProductBuilder.java [hunk 3]
```diff
@@ -239,6 +250,51 @@ public void startup() throws Exception {
...
+			sendExecutor.invokeAll(sendTasks, timeoutSeconds, TimeUnit.SECONDS);
```
[the remark this answers is not in the document]
Comment (jmfee-usgs) [reply src/main/java/gov/usgs/earthquake/distribution/CLIProductBuilder.java:284]: I thought about making them Callable and having it return null/Exception, but I think sendComplete and sendException capture the overall status where there is a timeout or another problem.

File: src/main/java/gov/usgs/earthquake/distribution/CLIProductBuilder.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (jmfee-usgs) [reply src/main/java/gov/usgs/earthquake/distribution/CLIProductBuilder.java:283]: Maybe, just generalize this Exception block so the client reports failure to send?

File: src/main/java/gov/usgs/earthquake/distribution/CLIProductBuilder.java [hunk 4]
```diff
@@ -239,6 +250,51 @@ public void startup() throws Exception {
...
+		} catch (InterruptedException e) {
+		}
+		sendExecutor.shutdown();
```
[the remark this answers is not in the document]
Comment (jmfee-usgs) [reply src/main/java/gov/usgs/earthquake/distribution/CLIProductBuilder.java:287]: Was using awaitTermination, but apparently that needs to occur after shutdown.  Also, invokeAll waits until tasks are complete, timeout, or are interrupted.

File: src/main/java/gov/usgs/earthquake/distribution/CLIProductBuilder.java [hunk 5]
```diff
@@ -239,6 +250,51 @@ public void startup() throws Exception {
...
+		// check whether send completed or was interrupted
+		for (ProductSender sender: sendComplete.keySet()) {
+			if (!sendComplete.get(sender) && sendExceptions.get(sender) == null) {
+				sendExceptions.put(sender, new InterruptedException());
```
[the remark this answers is not in the document]
Comment (jmfee-usgs) [reply src/main/java/gov/usgs/earthquake/distribution/CLIProductBuilder.java:291]: Yes, but the future's don't know which sender failed.  Maybe they are returned in the same order as the senders list, but this seemed easier to reason about.

File: src/main/java/gov/usgs/earthquake/distribution/CLIProductBuilder.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (emartinez-usgs) [reply src/main/java/gov/usgs/earthquake/distribution/CLIProductBuilder.java:285]: So what should we do about it?

File: src/main/java/gov/usgs/earthquake/distribution/CLIProductBuilder.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (emartinez-usgs) [reply src/main/java/gov/usgs/earthquake/distribution/CLIProductBuilder.java:283]: Sure, reporting send failures would be good I think.

File: src/main/java/gov/usgs/earthquake/distribution/CLIProductBuilder.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (emartinez-usgs) [reply src/main/java/gov/usgs/earthquake/distribution/CLIProductBuilder.java:291]: Yes, the returned futures is supposed to be iterable in the same order as the invoked tasks. This work too though.

File: src/main/java/gov/usgs/earthquake/distribution/CLIProductBuilder.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (jmfee-usgs) [reply src/main/java/gov/usgs/earthquake/distribution/CLIProductBuilder.java:283]: done

File: src/main/java/gov/usgs/earthquake/distribution/CLIProductBuilder.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (jmfee-usgs) [reply src/main/java/gov/usgs/earthquake/distribution/CLIProductBuilder.java]: Checking sendComplete and sendExceptions to report send status.  Added comment to that effect.

**Review by emartinez-usgs — CHANGES_REQUESTED** [review]
This seems rather disorganized to me.

Why is there a CLIProductBuilder that does not extend the normal ProductBuilder, but then there is an EIDSInputWedge that does extend the normal ProductBuilder, but then implicitly depends on the CLIProductBuilder (via static properties).

Curious why there are `PARALLEL_SEND_TIMEOUT_PROPERTY` as well as `PARALLEL_SEND_TIMEOUT_ARGUMENT`, but similar PROPRETY/ARGUMENT parallels don't exist for other configurations.

Maybe this all makes sense, but without explanation, it seems unclear.

**Comment by jmfee-usgs** [comment]
> This seems rather disorganized to me.
> 
> Why is there a CLIProductBuilder that does not extend the normal ProductBuilder, but then there is an EIDSInputWedge that does extend the normal ProductBuilder, but then implicitly depends on the CLIProductBuilder (via static properties).
> 
> Curious why there are `PARALLEL_SEND_TIMEOUT_PROPERTY` as well as `PARALLEL_SEND_TIMEOUT_ARGUMENT`, but similar PROPRETY/ARGUMENT parallels don't exist for other configurations.
> 
> Maybe this all makes sense, but without explanation, it seems unclear.

Agreed.  That seemed like a larger, riskier change at this point.  I added an issue https://github.com/usgs/pdl/issues/110 to refactor the CLIProductBuilder.   The properties probably aren't needed, but since the default was changing I wanted to give users an option to opt out.  Perhaps we can clean this all up for pdl 3.

**Pull request merged**